### ISSUE TITLE
(2023-2-6 10:00)  How to implement the gradient method of BroadcastTo in homework 1

### ISSUE THREAD

**Issue by asdfefbhjy** [OccupyMars2025/Deep-Learning-Systems-Algorithms-and-Implementation#3]
## Understand how an array or a tensor is stored on memory and how they are interpreted.

**Comment by asdfefbhjy** [comment]
```python
# import numpy as np


# x = np.array([1, 2, 3])
# x2 = np.broadcast_to(x, (3, 3))
# print(x)
# print(x2)
# print(id(x), id(x2), id(x) == id(x2), sep='\n')
# print(x is x2)
# print(20*"=")

# try:
#     x2[1, 0] = 99
#     print(x, x2, sep='\n')
# except ValueError as e:
#     print("Error!", e)
# print(30*"=")

# x[0] = 99
# print(x, x2, sep='\n')

import torch

x = torch.tensor([1, 2, 3])
x2 = torch.broadcast_to(x, (3, 3))
print(x, x2, sep='\n')
print(id(x), id(x2), id(x) == id(x2), sep='\n')
print(x is x2)
print(20*"=")

x[0] = 99
print(x, x2, sep='\n')
print(30*"=")

try:
    x2[1, 1] = 777
    print(x, x2, sep='\n')
except ValueError as e:
    print("Error!", e)
print(30*"=")

x = torch.tensor([1, 1, 1], dtype=torch.float32, requires_grad=True)
y = torch.broadcast_to(x, (3, 3))
out_gradient = torch.randn(3, 3)
out_gradient.requires_grad = True
print(out_gradient)
y.backward(out_gradient)
print(x)
print(x.grad)
print(torch.sum(out_gradient, 0))
```
# The output is as follows:
```shell
tensor([1, 2, 3])
tensor([[1, 2, 3],
        [1, 2, 3],
        [1, 2, 3]])
1805892326560
1805894405392
False
False
====================
tensor([99,  2,  3])
tensor([[99,  2,  3],
        [99,  2,  3],
        [99,  2,  3]])
==============================
tensor([ 99, 777,   3])
tensor([[ 99, 777,   3],
        [ 99, 777,   3],
        [ 99, 777,   3]])
==============================
tensor([[-0.6607, -0.0859,  0.6790],
        [ 0.3700, -0.8737, -0.5766],
        [-0.4892,  0.2963,  0.5612]], requires_grad=True)
tensor([1., 1., 1.], requires_grad=True)
tensor([-0.7800, -0.6634,  0.6636])
tensor([-0.7800, -0.6634,  0.6636], grad_fn=<SumBackward1>)

```
## You can see that the gradient method of `torch.broadcast_to` is  **to sum over the corresponding elements**

**Comment by asdfefbhjy** [comment]
# 2023-2-6 14:00  I implement the gradient method of BroadcastTo as follows: (It think it can be simplified)
```python
    def gradient(self, out_grad: Tensor, node: Tensor):
        original_shape = node.inputs[0].shape
        broadcast_shape = self.shape
        axes_to_sum_over = []
        for i in range(-1, -len(broadcast_shape) - 1, -1):
            if i >= -len(original_shape):
                if 1 == original_shape[i]:
                    axes_to_sum_over.append(i)
            else:
                axes_to_sum_over.append(i)
        # np.broadcast_to doesn't accept a list as the "axis" argument,
        # because now I use numpy as the backend, so I have to transform the list to a tuple
        axes_to_sum_over = tuple(axes_to_sum_over)
        
        temp = summation(out_grad, axes=axes_to_sum_over, keep_axes=False)
        return reshape(temp, shape=original_shape)
```
